cloudinit/distros/debian.py

```diff
@@ -155,7 +164,70 @@ def _get_localhost_ip(self):
     def set_timezone(self, tz):
         distros.set_etc_timezone(tz=tz, tz_file=self._find_tz_file(tz))
 
+    def _is_apt_lock_available(self, lock_files=APT_LOCK_FILES):
+        """Determines if another process holds any apt locks.
+
+        If all locks are clear, return True else False.
+        """
+        with ExitStack() as stack:
+            file_handles = []
+            for lock in lock_files:
+                try:
+                    file_handles.append(stack.enter_context(open(lock, 'w')))
+                except FileNotFoundError:
+                    # If the lock file doesn't even exist yet, then we don't
+                    # have to wait for it
+                    pass
+            for handle in file_handles:
+                try:
+                    fcntl.lockf(handle, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                except IOError:
+                    return False
+        return True
+
+    def _wait_for_apt_install(
+        self, short_cmd, subp_kwargs, timeout=APT_LOCK_WAIT_TIMEOUT
+    ):
+        """Wait for apt install to complete.
+
+        short_cmd: Name of command like "upgrade" or "install"
+        subp_kwargs: kwargs to pass to subp
+        """
+        start_time = time.time()
+        LOG.debug('Waiting for apt lock')
+        while time.time() - start_time < timeout:
+            if not self._is_apt_lock_available():
+                time.sleep(1)
+                continue
+            LOG.debug('apt lock clear')
+            try:
+                # Allow the output of this to flow outwards (not be captured)
+                log_msg = "apt-%s [%s]" % (
+                    short_cmd,
+                    ' '.join(subp_kwargs['args'])
+                )
+                return util.log_time(
+                    logfunc=LOG.debug,
+                    msg=log_msg,
+                    func=subp.subp,
+                    kwargs=subp_kwargs,
+                )
+            except subp.ProcessExecutionError as e:
+                if e.exit_code == 100 and 'Could not get lock' in e.stderr:
```

Comment:
Does this pose any internationalization concerns?

Reply:
A couple of concerns: 
 - Agree that hard-coding English strings to scope our logging of the exception reason is ineffective without using gettext.
 - I'm not sure this logging warrants us digging into gettext on the apt and apt-utils domain to do translations on this expected string type for just this log.
 - This loop looks to sleep and retry for 30 seconds on all CalledProcessErrors. Shouldn't we be raising and not retrying the called processerrors if exit is != 100 or if _is_apt_lock_available is True yet we received a CalledProcessError?

I'm not quite understanding what you are expecting to handle in this except clause:
- Are you thinking that our _is_apt_lock_available is not enough of a pre-flight check on obtaining- the locks or 
- Are you trying to handle a race where some other process obtained the APT lock between our success on _is_apt_lock_available and our subsequent call of subp?